### PR TITLE
feat: Async refresh dry run in parallel with sync refresh

### PR DESCRIPTION
**Pull request by bhshkh**
Internal bug : b/372244283

This is being used in Bigtable client 
https://github.com/googleapis/google-cloud-go/pull/11066

### CODE REVIEW

File: option/internaloption/internaloption.go [hunk 1]
```diff
@@ -209,6 +209,27 @@ func (w enableNewAuthLibrary) Apply(o *internal.DialSettings) {
 	o.EnableNewAuthLibrary = bool(w)
 }
 
+// EnableAsyncRefreshDryRun  returns a ClientOption that specifies if libraries in this
```
Comment (quartzmo) [inline option/internaloption/internaloption.go:212]: typo: extra space in "EnableAsyncRefreshDryRun  returns"

File: option/internaloption/internaloption.go [hunk 2]
```diff
@@ -209,6 +209,27 @@ func (w enableNewAuthLibrary) Apply(o *internal.DialSettings) {
...
+// module should asynchronously refresh auth token in parallel to sync refresh
+//
+// errHandler function will be called when there is an error while refreshing
+// the token asynchronously
+//
+// This is an experimental option and will be removed in the future
```
Comment (quartzmo) [inline option/internaloption/internaloption.go:218]: The three sentences above should all end with a period (`.`).

File: transport/grpc/dial.go [hunk 3]
```diff
@@ -262,6 +262,37 @@ func prepareDialOptsNewAuth(ds *internal.DialSettings) []grpc.DialOption {
 	return append(opts, ds.GRPCDialOpts...)
 }
 
+// dryRunAsync is a wrapper for oauth2.TokenSource that performs a sync refresh
+// after an async refresh. Token generated by async refresh is not used
```
Comment (quartzmo) [inline transport/grpc/dial.go:266]: Missing period (`.`) after "used".

File: option/internaloption/internaloption.go [hunk 4]
```diff
@@ -209,6 +209,27 @@ func (w enableNewAuthLibrary) Apply(o *internal.DialSettings) {
 	o.EnableNewAuthLibrary = bool(w)
 }
 
+// EnableAsyncRefreshDryRun  returns a ClientOption that specifies if libraries in this
+// module should asynchronously refresh auth token in parallel to sync refresh
```
Comment (quartzmo) [inline option/internaloption/internaloption.go:213]: Can you add more information here explaining why it is a "dry run" and what the goal or purpose is?

File: option/internaloption/internaloption.go [hunk 5]
```diff
@@ -209,6 +209,27 @@ func (w enableNewAuthLibrary) Apply(o *internal.DialSettings) {
...
+//
+// errHandler function will be called when there is an error while refreshing
+// the token asynchronously
```
Comment (quartzmo) [inline option/internaloption/internaloption.go:216]: What should the errHandler function do? Log the error?

File: option/internaloption/internaloption.go (hunk 2 above)
Comment (quartzmo) [reply option/internaloption/internaloption.go:218]: I think "experimental" should be in all caps: "EXPERIMENTAL".

File: internal/settings.go [hunk 6]
```diff
@@ -70,8 +70,9 @@ type DialSettings struct {
 	RequestReason string
 
 	// New Auth library Options
-	AuthCredentials      *auth.Credentials
-	EnableNewAuthLibrary bool
+	AuthCredentials          *auth.Credentials
+	EnableNewAuthLibrary     bool
+	EnableAsyncRefreshDryRun func()
```
Comment (quartzmo) [inline internal/settings.go:75]: Should all of the changes in this PR be marked with `TODO` comments explaining that they are temporary and should be removed after some event?

**Review by quartzmo — CHANGES_REQUESTED** [review]

File: option/internaloption/internaloption.go (hunk 5 above)
Comment (bhshkh) [reply option/internaloption/internaloption.go]: No specific requirement on that. User can do whatever they need to

File: option/internaloption/internaloption.go (hunk 1 above)
Comment (bhshkh) [reply option/internaloption/internaloption.go]: Done

File: option/internaloption/internaloption.go (hunk 2 above)
Comment (bhshkh) [reply option/internaloption/internaloption.go]: Done

File: option/internaloption/internaloption.go (hunk 4 above)
Comment (bhshkh) [reply option/internaloption/internaloption.go]: Done

File: option/internaloption/internaloption.go (hunk 5 above)
Comment (bhshkh) [reply option/internaloption/internaloption.go]: Done

File: transport/grpc/dial.go (hunk 3 above)
Comment (bhshkh) [reply transport/grpc/dial.go]: Done

**Comment by codyoss** [comment]
Can this change be made upstream instead with an embedded private option: https://pkg.go.dev/google.golang.org/api/option/internaloption#EmbeddableAdapter

**Review by quartzmo — APPROVED** [review]

**Comment by bhshkh** [comment]
> Can this change be made upstream instead with an embedded private option: https://pkg.go.dev/google.golang.org/api/option/internaloption#EmbeddableAdapter

No. the token refresh code is deep into this library and not Bigtable.

File: option/internaloption/internaloption.go [hunk 7]
```diff
@@ -211,6 +211,33 @@ func (w enableNewAuthLibrary) Apply(o *internal.DialSettings) {
...
+//
+// This option can be used to determine whether refreshing the token asymnchronously
+// prior to its actual expiry works without any issues in a particular environment.
+//
+// errHandler function will be called when there is an error while refreshing
+// the token asynchronously.
+//
+// This is an EXPERIMENTAL option and will be removed in the future.
+// TODO(b/372244283): Remove after b/358175516 has been fixed
+func EnableAsyncRefreshDryRun(errHandler func()) option.ClientOption {
```
Comment (mutianf) [inline option/internaloption/internaloption.go:225]: Is there any annotation to mark this as internal api?

File: transport/grpc/dial.go [hunk 8]
```diff
@@ -298,8 +332,14 @@ func dial(ctx context.Context, insecure bool, o *internal.DialSettings) (*grpc.C
 			if err != nil {
 				return nil, err
 			}
+
+			ts := creds.TokenSource
```
Comment (codyoss) [inline transport/grpc/dial.go:336]: This code will not be hit if the new auth lib is enabled. That that intentional here?

Comment (bhshkh) [reply transport/grpc/dial.go:336]: Yes

File: option/internaloption/internaloption.go (hunk 7 above)
Comment (bhshkh) [reply option/internaloption/internaloption.go:225]: No.

**Comment by bhshkh** [comment]
@codyoss I don't have permission to merge this. Can you please merge it if there aren't anymore comments

**Pull request merged**